x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/LatLonShapeDocValuesQuery.java

```diff
@@ -0,0 +1,193 @@
+        if (field == null) {
+            throw new IllegalArgumentException("field must not be null");
+        }
+        this.field = field;
+        this.geometries = geometries;
+        this.relation = relation;
+    }
+
+    @Override
+    public String toString(String field) {
+        StringBuilder sb = new StringBuilder();
+        if (!this.field.equals(field)) {
+            sb.append(this.field);
+            sb.append(':');
+            sb.append(relation);
+            sb.append(':');
+        }
+        sb.append(Arrays.toString(geometries));
+        return sb.toString();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (sameClassAs(obj) == false) {
+            return false;
+        }
+        LatLonShapeDocValuesQuery other = (LatLonShapeDocValuesQuery) obj;
+        return field.equals(other.field) && relation == other.relation && Arrays.equals(geometries, other.geometries);
+    }
+
+    @Override
+    public int hashCode() {
+        int h = classHash();
+        h = 31 * h + field.hashCode();
+        h = 31 * h + relation.hashCode();
+        h = 31 * h + Arrays.hashCode(geometries);
+        return h;
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+        if (visitor.acceptField(field)) {
+            visitor.visitLeaf(this);
+        }
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) {
+         if (relation == ShapeField.QueryRelation.CONTAINS) {
+             return getContainsWeight(scoreMode, boost);
+         } else {
+            return getStandardWeight(scoreMode, boost);
+         }
+    }
+
+    private ConstantScoreWeight getStandardWeight(ScoreMode scoreMode, float boost) {
+        return new ConstantScoreWeight(this, boost) {
+            final Component2D component2D = LatLonGeometry.create(geometries);
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                final BinaryDocValues values = context.reader().getBinaryDocValues(field);
+                if (values == null) {
+                    return null;
+                }
+                final GeometryDocValueReader reader = new GeometryDocValueReader();
+                final Component2DVisitor visitor = Component2DVisitor.getVisitor(component2D, relation, CoordinateEncoder.GEO);
+
+                final TwoPhaseIterator iterator = new TwoPhaseIterator(values) {
+
+                    @Override
+                    public boolean matches() throws IOException {
+                        reader.reset(values.binaryValue());
+                        visitor.reset();
+                        reader.visit(visitor);
+                        return visitor.matches();
+                    }
+
+                    @Override
+                    public float matchCost() {
+                        return 1000f; // TODO: what should it be?
```

Comment:
I am not too familiar with this. So long as `matchCost` is constant and the ConstantScoreScorer is used, this number is arbitrary and has no effect, right?

Reply:
My understanding is that this value is used when you have more that one `TwoPhaseIterator` in a conjunction in order to sort them. You want to execute the cheaper iterators first.

In put case we do not have a good estimation of cost so it is good enough to return a big number. Note that this is done the same for LatLonPointDocValues.

Reply:
thanks!